File: leetcode/654-最大二叉树/654-最大二叉树.go
package main

import "math"

/**
* Definition for a binary tree node.

 */
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 654-最大二叉树
func constructMaximumBinaryTree(nums []int) *TreeNode {
	if nums == nil || len(nums) == 0 {
		return nil
	}
	return buildTree(nums, 0, len(nums)-1)
}

func buildTree(nums []int, l, r int) *TreeNode {
	if l > r {
		return nil
	}
	// 寻找最大值，记录索引
	index := -1
	maxVal := math.MinInt
	for i := l; i <= r; i++ {
		if nums[i] > maxVal {
			index = i
			maxVal = nums[i]
		}
	}
	node := &TreeNode{Val: maxVal}
	node.Left = buildTree(nums, l, index-1)
	node.Right = buildTree(nums, index+1, r)
	return node
}

File: leetcode/654-最大二叉树/654-最大二叉树.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

# 654-最大二叉树
class Solution:
    def constructMaximumBinaryTree(self, nums: List[int]) -> Optional[TreeNode]:
        if len(nums) == 0:
             return None
        def build_tree(nums,start,end):
            if start > end:
                return None
            index,max_val = 0,-sys.maxsize - 1
            for i in range(start,end+1):
                if nums[i] > max_val:
                    index = i
                    max_val = nums[i]
            root = TreeNode(max_val)
            root.left = build_tree(nums,start,index-1)
            root.right = build_tree(nums,index+1,end)
            return root
        
        return build_tree(nums,0,len(nums)-1)